### ISSUE TITLE
autocompletion does not work if array given for an object

### ISSUE THREAD

**Issue by gencer** [castwide/solargraph#106]
When I declare type for solargraph and fetch an object from an array, autocompletion does not work. See here:

![image](https://user-images.githubusercontent.com/1711576/48670786-ce7b2500-eb2e-11e8-8b7f-f9c588ab6150.png)

If I do this by giving a method or a static member then it works:

![image](https://user-images.githubusercontent.com/1711576/48670788-f7031f00-eb2e-11e8-84ba-526f94e0c141.png)

Basically, `anything[]` type does not support autocomplete. I cannot see methods in it.

Sample:
```
    # @type [Reserve]
    source = Reserve[source_id]
    # @type [Reserve]
    another = anonymous_method()
    
    source. # does not work.
    another. # works.
```

**Comment by gencer** [comment]
@castwide, I'm not sure if this and above related to this gem or vscode extension but here is another issue:

![first](https://user-images.githubusercontent.com/1711576/48670904-3fbbd780-eb31-11e8-819b-91994cfa06e6.gif)

**Comment by castwide** [comment]
In your first screenshot, it looks like `def perform` lacks an `end`. Solargraph will attempt to correct for parser errors while the file's being edited, but is sometimes unable to do so. Here are my results with and without a closing `end`:

![image](https://user-images.githubusercontent.com/3704922/48672872-81d12300-eb09-11e8-8ab8-40514dd257bf.png)

![image](https://user-images.githubusercontent.com/3704922/48672874-885f9a80-eb09-11e8-981b-ce646fcef210.png)

If your second screenshot isn't inside an unclosed method definition, that could explain why it works.

The third screenshot might also be due to a similar parser error (`if` without `end`), although Solargraph should be able to detect and correct it in most cases:

![image](https://user-images.githubusercontent.com/3704922/48672920-0a4fc380-eb0a-11e8-8391-a574edbb88e3.png)

**Comment by gencer** [comment]
It is interesting that it work for you.

I had `end`'s already. But i figured it out why it is not working. I had a `rescue` at bottom. See this GIF:

![first](https://user-images.githubusercontent.com/1711576/48673138-5b35db00-eb4f-11e8-8710-4488d527cfdd.gif)

As you can see, If I have a `rescue`, `if` and `source.` does not work. If I remove `rescue` part it just works both for `if` and `source.`

**Comment by castwide** [comment]
Thanks for the followup. I was able to reproduce the problem with `rescue`. I'll work on a fix.

**Comment by castwide** [comment]
Fortunately, the new source mapper made this relatively simple to fix. `rescue` nodes work in the master branch. There was a related issue with explicit `begin` nodes which is also fixed. I'll include the changes in gem version 0.29.0.

**Comment by gencer** [comment]
I'm happy to hear that and await for the new version.

Thanks!

**Comment by gencer** [comment]
@castwide, does this fix currently in `master`?

**Comment by gencer** [comment]
@castwide Just wanted to be sure. Here is another issue:

If you assign a variable and tell solargraph of its type but use with a condition, autocomplete does not work.

![first](https://user-images.githubusercontent.com/1711576/48953938-99206e00-ef58-11e8-85ad-5a82a5c463a4.gif)


Basically:

```
# @type [Array<Object>]
a = b
a. # <- auto complete works and Object get listed.

---

# @type [Array<Object>]
a = b unless b.nil?
a. # <- autocomplete does not work because we used unless/if.

```

**Comment by castwide** [comment]
Confirmed. It's a minor bug that's also related to the new source mapper. I'll have it fixed in master shortly and will include the fix in v0.29.0.

**Comment by gencer** [comment]
@castwide one more question. Is there any way to use YARD documentation for `attr_`'s?

For example:

```
class User
   # @type [Comment]
   attr_reader :comment
end

user = User.new
user.comment. # <- I expect Comment object methods here but it does not work.
```

**Comment by gencer** [comment]
@castwide, Gotcha. Sorry for this. 

I also send you an e-mail. Sample video + Sample project. There is an issue with autocomplete when nested classes used. You can watch there and test yourself.

Gencer.

**Comment by castwide** [comment]
@gencer The bug related to your last email is being tracked in #108. The fix is scheduled for release in v0.29.0.

**Comment by castwide** [comment]
Fix released in v0.29.0.

**Comment by gencer** [comment]
@castwide, This issue rise from the dead again. See here:

![first](https://user-images.githubusercontent.com/1711576/50426140-bb7b1380-0897-11e9-9c52-d3cd24c9f05e.gif)

**Comment by gencer** [comment]
@castwide, if you get an e-mail about this issue, please ignore it. My bad.
